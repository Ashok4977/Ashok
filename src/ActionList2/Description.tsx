import React from 'react'
import Box from '../Box'
import Truncate from '../Truncate'
import {Slot} from './Item'

export type DescriptionProps = {
  variant?: 'inline' | 'block'
}
export const Description: React.FC<DescriptionProps> = ({variant = 'inline', ...props}) => {
  const styles = {
    color: 'fg.muted',
    fontSize: 0,
    lineHeight: '16px',
    flexGrow: 1,
    flexBasis: 0,
    minWidth: 0,
    marginLeft: variant === 'block' ? 0 : 2
  }

  return (
    <Slot name={variant === 'block' ? 'BlockDescription' : 'InlineDescription'}>
      {variant === 'block' ? (
        <Box as="span" sx={styles}>
          {props.children}
        </Box>
      ) : (
        <Truncate sx={styles} title={props.children as string} inline={true} maxWidth="100%">
          {props.children}
        </Truncate>
      )}
    </Slot>
  )
}
