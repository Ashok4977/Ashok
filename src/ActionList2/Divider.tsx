import React from 'react'
import Box from '../Box'
import {get} from '../constants'
import {Theme} from '../ThemeProvider'

/**
 * Visually separates `Item`s or `Group`s in an `ActionList`.
 */
export function Divider(): JSX.Element {
  return (
    <Box
      as="hr"
      sx={{
        border: 'none', // override browser styles
        height: 1,
        backgroundColor: 'border.muted',
        marginTop: (theme: Theme) => `calc(${get('space.2')(theme)} - 1px)`,
        marginBottom: 2
      }}
      data-component="ActionList.Divider"
    />
  )
}

/**
 * `Divider` fulfills the `ItemPropsWithCustomRenderer` contract,
 * so it can be used inline in an `ActionList`â€™s `items` prop.
 * In other words, `items={[ActionList.Divider]}` is supported as a concise
 * alternative to `items={[{renderItem: () => <ActionList.Divider />}]}`.
 */
Divider.renderItem = Divider
