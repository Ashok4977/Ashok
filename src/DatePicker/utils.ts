import deepmerge from 'deepmerge'
import {DatePickerConfiguration, isMultiSelection, isRangeSelection, Selection} from './types'
import {defaultConfiguration} from './DatePickerProvider'

export const sanitizeDate = (date: Date | string) => new Date(new Date(date).toDateString())

export const getInitialFocusDate = (selection?: Selection) => {
  if (!selection) return sanitizeDate(new Date())

  if (selection instanceof Date) {
    return sanitizeDate(selection)
  } else if (isMultiSelection(selection)) {
    return sanitizeDate(selection[0])
  } else if (isRangeSelection(selection)) {
    return sanitizeDate(selection.from)
  } else {
    return sanitizeDate(new Date())
  }
}

export const initializeConfigurations = (externalConfig?: DatePickerConfiguration) => {
  let sanitizedConfig = {}
  if (externalConfig) {
    sanitizedConfig = Object.entries(externalConfig).filter(item => item[1] !== undefined && item[1] !== null)
    for (const item of sanitizedConfig) {
      if (item instanceof Date) {
        sanitizeDate(item)
      }
    }
  }
  return deepmerge(defaultConfiguration, sanitizedConfig)
}
