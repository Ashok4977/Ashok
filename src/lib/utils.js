
/**
 * Splice items from an Array for with the test function returns
 * truthy, modifying the original list and returning an array of
 * the spliced items.
 *
 * ```js
 * const children = React.Children.toArray(props.children)
 * const buttons = splice(children, child => child.type === Button)
 * ```
 */
const splice = (list, test) => {
  return list.reduce((acc, item, i) => {
    if (test(item)) {
      return acc.concat(list.splice(i, 1))
    }
    return acc
  }, [])
}

export {splice}
