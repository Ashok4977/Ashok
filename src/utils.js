import chroma from 'chroma-js'
import {css} from 'styled-components'
import {theme} from 'styled-system'

// color shortcuts
export const color = name => theme(`colors.${name}`)
export const black = color('black')
export const white = color('white')

/**
 * Reduce a value to a primitive as long as it is a function.
 *
 * interpolate(() => 'x') // returns 'x'
 * interpolate(() => () => 'y') // returns 'y'
 */
export const interpolate = (value, context, props) => {
  while (typeof value === 'function') {
    value = value.call(context, props)
  }
  return value
}

/**
 * This decorator interpolates all of the function's arguments by calling any
 * functions in the provided context and with the provided arguments. In other
 * words, the function you provide is guaranteed to have all of its arguments
 * "called" if they are functions. Example:
 *
 * const rgba = interpolator((rgb, a) => chroma(rgb).alpha(a).css())
 * const fadeed = rgba(theme('colors.black'), .5) // returns a function!
 * faded({}) // probably throws an error because props.theme isn't set
 */
export const interpolator = f => {
  return (...outer) => {
    return function(props) {
      return f(...[...outer].map(a => interpolate(a, this, props)))
    }
  }
}

export const rgba = interpolator(       (rgb, alpha)    => chroma(rgb).alpha(alpha).css())
export const lighten = interpolator(    (value, amount) => chroma(value).brighten(amount).css())
export const darken = interpolator(     (value, amount) => chroma(value).darken(amount).css())
export const desaturate = interpolator( (value, amount) => chroma(value).desaturate(amount).css())
export const mix = interpolator(        (a, b, amount)  => chroma.mix(a, b, amount).css())
