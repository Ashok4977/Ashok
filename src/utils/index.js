
export function defined(val) {
  return val !== null && typeof val !== 'undefined'
}

export function definedFallback(input, fallback) {
  return defined(input) ? input : fallback
}
