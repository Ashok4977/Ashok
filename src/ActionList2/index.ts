import {List} from './List'
import {Group} from './Group'
import {Item} from './Item'
import {Divider} from './Divider'
import {Description} from './Description'
import {LeadingVisual, TrailingVisual} from './Visuals'

export type {ListProps as ActionListProps} from './List'
export type {GroupProps} from './Group'
export type {ItemProps} from './Item'
export type {DescriptionProps} from './Description'
export type {LeadingVisualProps, TrailingVisualProps} from './Visuals'

/**
 * Collection of list-related components.
 */
export const ActionList = Object.assign(List, {
  /** Collects related `Items` in an `ActionList`. */
  Group,

  /** An actionable or selectable `Item` with an optional icon and description. */
  Item,

  /** Visually separates `Item`s or `Group`s in an `ActionList`. */
  Divider,

  Description,

  LeadingVisual,
  TrailingVisual
})
