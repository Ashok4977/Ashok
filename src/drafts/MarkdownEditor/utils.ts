import {isMacOS} from '@primer/behaviors/utils'

export type InputState = {
  readonly value: string
  readonly selectionStart: number
  readonly selectionEnd: number
}

export const getSelectedLineRange = (inputState: InputState): [number, number] => {
  // Subtract one from the caret position so the newline found is not the one _at_ the caret position
  // then add one because we don't want to include the found newline. Also changes -1 (not found) result to 0
  const start = inputState.value.lastIndexOf('\n', inputState.selectionStart - 1) + 1

  // activeLineEnd will be the index of the next newline inclusive, which works because slice is last-index exclusive
  let end = inputState.value.indexOf('\n', inputState.selectionEnd)
  if (end === -1) end = inputState.value.length

  return [start, end]
}

export const markdownComment = (text: string) => `<!-- ${text.replaceAll('--', '\\-\\-')} -->`

export const markdownLink = (text: string, url: string) =>
  `[${text.replaceAll('[', '\\[').replaceAll(']', '\\]')}](${url.replaceAll('(', '\\(').replaceAll(')', '\\)')})`

export const markdownImage = (altText: string, url: string) => `!${markdownLink(altText, url)}`

export const isModifierKey = (event: KeyboardEvent | React.KeyboardEvent<unknown>) =>
  isMacOS() ? event.metaKey : event.ctrlKey

export function isMultipleLines(string: string): boolean {
  return string.trim().split('\n').length > 1
}
