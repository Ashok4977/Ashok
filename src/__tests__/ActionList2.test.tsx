import {cleanup, render as HTMLRender, waitFor, fireEvent} from '@testing-library/react'
import 'babel-polyfill'
import {axe, toHaveNoViolations} from 'jest-axe'
import React from 'react'
import theme from '../theme'
import {ActionList} from '../ActionList2'
import {behavesAsComponent, checkExports, checkStoriesForAxeViolations} from '../utils/testing'
import {BaseStyles, ThemeProvider, SSRProvider} from '..'
import '@testing-library/jest-dom'
expect.extend(toHaveNoViolations)

function SimpleActionList(): JSX.Element {
  return (
    <ThemeProvider theme={theme}>
      <SSRProvider>
        <BaseStyles>
          <ActionList>
            <ActionList.Item>New file</ActionList.Item>
            <ActionList.Divider />
            <ActionList.Item>Copy link</ActionList.Item>
            <ActionList.Item>Edit file</ActionList.Item>
            <ActionList.Item variant="danger">Delete file</ActionList.Item>
          </ActionList>
        </BaseStyles>
      </SSRProvider>
    </ThemeProvider>
  )
}

const projects = [
  {name: 'Primer Backlog', scope: 'GitHub'},
  {name: 'Primer React', scope: 'github/primer'}
]
function SingleSelectListStory(): JSX.Element {
  const [selectedIndex, setSelectedIndex] = React.useState(0)

  return (
    <ActionList selectionVariant="single" showDividers role="listbox" aria-label="Select a project">
      {projects.map((project, index) => (
        <ActionList.Item
          key={index}
          role="option"
          selected={index === selectedIndex}
          onSelect={() => setSelectedIndex(index)}
        >
          {project.name}
        </ActionList.Item>
      ))}
    </ActionList>
  )
}

describe('ActionList', () => {
  behavesAsComponent({
    Component: ActionList,
    options: {skipAs: true, skipSx: true},
    toRender: () => <ActionList />
  })

  checkExports('ActionList2', {
    default: undefined,
    ActionList
  })

  it('should have no axe violations', async () => {
    const {container} = HTMLRender(<SimpleActionList />)
    const results = await axe(container)
    expect(results).toHaveNoViolations()
    cleanup()
  })

  it('should fire onSelect', async () => {
    const component = HTMLRender(<SingleSelectListStory />)
    const options = await waitFor(() => component.getAllByRole('option'))

    expect(options[0]).toHaveAttribute('aria-selected', 'true')
    expect(options[1]).toHaveAttribute('aria-selected', 'false')

    fireEvent.click(options[1])

    expect(options[0]).toHaveAttribute('aria-selected', 'false')
    expect(options[1]).toHaveAttribute('aria-selected', 'true')

    cleanup()
  })

  checkStoriesForAxeViolations('ActionList2')
})
