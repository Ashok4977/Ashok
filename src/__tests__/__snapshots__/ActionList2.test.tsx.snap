// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionList renders consistently 1`] = `
.c0 {
  margin: 0;
  padding-inline-start: 0;
  padding-top: 8px;
  padding-bottom: 8px;
}

<ul
  className="c0"
/>
`;

exports[`ActionList.Item renders consistently 1`] = `
.c4:hover:not([aria-disabled]) + .c0,
.c4:focus:not([aria-disabled]) + .c4 {
  --divider-color: transparent;
}

.c4[aria-selected=true] + .c0 {
  --divider-color: transparent;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  min-width: 0;
}

.c3 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c5:hover:not([aria-disabled]) + .c0,
.c5:focus:not([aria-disabled]) + .c5 {
  --divider-color: transparent;
}

.c5[aria-selected=true] + .c0 {
  --divider-color: transparent;
}

.c6:hover:not([aria-disabled]) + .c0,
.c6:focus:not([aria-disabled]) + .c6 {
  --divider-color: transparent;
}

.c6[aria-selected=true] + .c0 {
  --divider-color: transparent;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 8px;
  padding-right: 8px;
  font-size: 14px;
  padding-top: 6px;
  padding-bottom: 6px;
  line-height: 20px;
  margin-left: 0;
  margin-right: 0;
  min-height: 5px;
  border-radius: 6px;
  -webkit-transition: background 33.333ms linear;
  transition: background 33.333ms linear;
  color: #24292f;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c1[aria-selected=true] {
  background-color: rgba(208,215,222,0.24);
}

.c1:not([aria-disabled]) {
  cursor: pointer;
}

.c1 [data-component="ActionList.Item--DividerContainer"] {
  position: relative;
}

.c1 [data-component="ActionList.Item--DividerContainer"]::before {
  content: " ";
  display: block;
  position: absolute;
  width: 100%;
  top: -7px;
  border: 0 solid;
  border-top-width: 0;
  border-color: var(--divider-color,transparent);
}

.c1:not(:first-of-type):not([aria-selected=true]) {
  --divider-color: rgba(208,215,222,0.48);
}

[data-component="ActionList.Divider"] + .c1 {
  --divider-color: transparent !important;
}

.c1:hover:not([aria-disabled]),
.c1:focus:not([aria-disabled]) {
  --divider-color: transparent;
}

.c1:hover:not([aria-disabled]) + .c0,
.c1:focus:not([aria-disabled]) + .c1 {
  --divider-color: transparent;
}

.c1[aria-selected=true] + .c0 {
  --divider-color: transparent;
}

@media (hover:hover) and (pointer:fine) {
  .c1:hover:not([aria-disabled]) {
    background-color: rgba(208,215,222,0.32);
    color: #24292f;
  }

  .c1:focus:not([aria-disabled]) {
    background-color: rgba(208,215,222,0.48);
    color: #24292f;
    outline: none;
  }
}

<li
  aria-describedby=""
  aria-labelledby="react-aria2220704418-19"
  className="c0 c1"
  onClick={[Function]}
  tabIndex={-1}
>
  <div
    className="c0 c2"
    data-component="ActionList.Item--DividerContainer"
  >
    <span
      className="c0 c3"
      id="react-aria2220704418-19"
    />
  </div>
</li>
`;
